compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -175,23 +204,42 @@ pub enum TyKind<'tcx> {
     Projection(ProjectionTy<'tcx>),
 
     /// Opaque (`impl Trait`) type found in a return type.
+    ///
     /// The `DefId` comes either from
     /// * the `impl Trait` ast::Ty node,
     /// * or the `type Foo = impl Trait` declaration
-    /// The substitutions are for the generics of the function in question.
-    /// After typeck, the concrete type can be found in the `types` map.
+    ///
+    /// For RTIT the substitutions are for the generics of the function,
+    /// while for TAIT it is used for the generic parameters of the alias.
     Opaque(DefId, SubstsRef<'tcx>),
 
     /// A type parameter; for example, `T` in `fn f<T>(x: T) {}`.
     Param(ParamTy),
 
-    /// Bound type variable, used only when preparing a trait query.
+    /// Bound type variable, used to represent the `'a` in `for<'a> fn(&'a ())`.
+    ///
+    /// For canonical queries, we replace inference variables with bound variables,
+    /// so e.g. when checking whether `&'_ (): Trait<_>` holds, we canonicalize that to
+    /// `for<'a, T> &'a (): Trait<T>` and then convert the introduced bound variables
+    /// back to inference variables in a new inference context when inside of the query.
+    ///
+    /// See the `rustc-dev-guide` for more details about
+    /// [higher-ranked trait bounds][1] and [canonical queries][2].
+    ///
+    /// [1]: https://rustc-dev-guide.rust-lang.org/traits/hrtb.html
+    /// [2]: https://rustc-dev-guide.rust-lang.org/traits/canonical-queries.html
     Bound(ty::DebruijnIndex, BoundTy),
 
-    /// A placeholder type - universally quantified higher-ranked type.
+    /// A placeholder type, used during higher ranked subtyping to instantiate
+    /// bound variables.
```

Comment:
is that all placeholders are used for?

Reply:
I think they are also used while traversing query boundaries (inference vars are converted to them, then the query is run, inside the query they are turned back into new inference vars. When returning a value from the query, inference vars are again turned into placeholders, once the query has returned, those placeholders are turned back into to the original inference vars)

Reply:
> think they are also used while traversing query boundaries (inference vars are converted to them, then the query is run, inside the query they are turned back into new inference vars.

don't we convert them to `ty::Bound`?

> Currently they are used for that, they represent "some unknown type".

"that" being "used during higher ranked subtyping to instantiate bound variables"?